CMakeLists.txt

```diff
@@ -0,0 +1,270 @@
+    ${CMAKE_CURRENT_SOURCE_DIR}/demo/test.c
+)
+
+target_include_directories(test-target PRIVATE
+    ${CMAKE_CURRENT_SOURCE_DIR}
+)
+
+target_link_libraries(test-target PRIVATE
+    ${PROJECT_NAME}
+)
+
+target_compile_options(test-target BEFORE PRIVATE
+    $<$<STREQUAL:$<TARGET_PROPERTY:${PROJECT_NAME},TYPE>,SHARED_LIBRARY>:-O1 -DLTM_TEST_DYNAMIC>
+    ${LTM_C_FLAGS}
+)
+target_link_options(test-target BEFORE PRIVATE
+    $<$<STREQUAL:$<TARGET_PROPERTY:${PROJECT_NAME},TYPE>,SHARED_LIBRARY>:-O1>
+    ${LTM_LD_FLAGS}
+)
+
+#-----------------------------------------------------------------------------
+# demo target
+#-----------------------------------------------------------------------------
+add_custom_target(check
+    COMMAND ${CMAKE_CURRENT_BINARY_DIR}/test-target
+)
+
+add_dependencies(check test-target)
+
+#-----------------------------------------------------------------------------
+# Install/export targets and files
+#-----------------------------------------------------------------------------
+set(CONFIG_INSTALL_DIR "${CMAKE_INSTALL_LIBDIR}/cmake/${PROJECT_NAME}")
+set(PROJECT_VERSION_FILE "${CMAKE_CURRENT_BINARY_DIR}/${PROJECT_NAME}-config-version.cmake")
+set(PROJECT_CONFIG_FILE "${PROJECT_NAME}-config.cmake")
+set(TARGETS_EXPORT_NAME "${PROJECT_NAME}Targets")
+
+install(TARGETS ${PROJECT_NAME}
+    EXPORT ${TARGETS_EXPORT_NAME}
+    LIBRARY DESTINATION ${CMAKE_INSTALL_LIBDIR}
+    ARCHIVE DESTINATION ${CMAKE_INSTALL_LIBDIR}
+)
+
+install(FILES ${HEADERS}
+    DESTINATION ${CMAKE_INSTALL_INCLUDEDIR}/${PROJECT_NAME}
+)
+
+# Install libtommath.pc for pkg-config if we build a shared library
+if(BUILD_SHARED_LIBS)
+    configure_file(
+        ${CMAKE_CURRENT_SOURCE_DIR}/lib${PROJECT_NAME}.pc.in
+        ${CMAKE_CURRENT_BINARY_DIR}/lib${PROJECT_NAME}.pc
+        @ONLY
+    )
+
+    install(FILES ${CMAKE_CURRENT_BINARY_DIR}/lib${PROJECT_NAME}.pc
+        DESTINATION ${CMAKE_INSTALL_DATAROOTDIR}/pkgconfig
+    )
+endif()
+
+# generate package version file
+write_basic_package_version_file(
+    ${PROJECT_VERSION_FILE}
+    VERSION ${PROJECT_VERSION}
+    COMPATIBILITY SameMajorVersion
+)
+
+# install version file
+install(FILES ${PROJECT_VERSION_FILE}
+        DESTINATION ${CONFIG_INSTALL_DIR}
+)
+
+# build directory package config
+export(EXPORT ${TARGETS_EXPORT_NAME}
+       FILE ${PROJECT_CONFIG_FILE}
+)
+
+# installed package config
+install(EXPORT ${TARGETS_EXPORT_NAME}
+        DESTINATION ${CONFIG_INSTALL_DIR}
+        FILE ${PROJECT_CONFIG_FILE}
+)
+
+# add to CMake registry
+export(PACKAGE ${PROJECT_NAME})
+
+#---------------------------------------------------------------------------------------
+# Create release packages
+#---------------------------------------------------------------------------------------
+
+# determine distribution and architecture
+find_program(LSB_RELEASE lsb_release)
+
+execute_process(COMMAND uname -m OUTPUT_VARIABLE MACHINE_ARCH OUTPUT_STRIP_TRAILING_WHITESPACE)
```

Comment:
You're probably right. I'd rather use [CMAKE_HOST_SYSTEM_NAME](https://cmake.org/cmake/help/v3.10/variable/CMAKE_HOST_SYSTEM_NAME.html#cmake-host-system-name) though.

Reply:
`CMAKE_HOST_SYSTEM_NAME` is the name of the system running cmake (the host system for cmake).
`CMAKE_SYSTEM_NAME` is the name of the system we are building for.

So `CMAKE_HOST_SYSTEM_NAME` is the build system name,
and `CMAKE_SYSTEM_NAME` is the host system name.

(The naming is counter-intuitive)

Reply:
They hold host and target. A simple CMakeLists.txt just printing these variables:

Building on linux for linux:
```
cmake ..
...
CMAKE_HOST_SYSTEM_NAME: Linux
CMAKE_SYSTEM_NAME     : Linux
...
```

Building on linux for windows:
```
x86_64-w64-mingw32-cmake ..
...
CMAKE_HOST_SYSTEM_NAME: Linux
CMAKE_SYSTEM_NAME     : Windows
...
```

Reply:
See https://crosstool-ng.github.io/docs/toolchain-types/
There exists build, host and target.
Since libtommath is not a compiler, host == target.

Reply:
The crosstool link may be a little confusing since it throws toolchain building in the mix.

We are just running a toolchain to build the lib.

So we are left with two questions:

1. In what system is the toolchain running i.e. the system where the lib is built (host)
2. In what system is the built lib going to run (target)

Bullet point 1. from the link matches the first example above:
    Build (toolchain) runs under linux and produces a lib to be run on linux.
    Thus host=linux target=linux

Bullet point 2.  from the link matches the second one:
    Build (toolchain) runs under linux and produces a lib to be run on windows.
    Thus host=linux target=windows

So cmake variable CMAKE_HOST_SYSTEM_NAME is the host and CMAKE_SYSTEM_NAME is the target.

Hope this clarifies it.

Reply:
> So cmake variable CMAKE_HOST_SYSTEM_NAME is the host and CMAKE_SYSTEM_NAME is the target and they are not necessarily the same.

:+1: 

In those parts of the file we're defining how the cpack file is called, so the name shall be derived from `CMAKE_SYSTEM_NAME`.

btw. using regular CMake with the toolchain file provided by [`mingw-w64-cmake`](https://aur.archlinux.org/packages/mingw-w64-cmake) produces a functional binary.

```
$ cmake --toolchain=/usr/share/mingw/toolchain-x86_64-w64-mingw32.cmake ..
...
-- CMAKE_HOST_SYSTEM_NAME="Linux"
-- CMAKE_SYSTEM_NAME="Windows"
-- CMAKE_EXECUTABLE_SUFFIX=".exe"
-- MINGW="1"
...
$ make -j65 check
...
Tests OK/NOP/FAIL: 47/0/0
[100%] Built target check
$ file test-target*
test-target.exe: PE32+ executable (console) x86-64, for MS Windows
```

Using [`mingw-w64-cmake`](https://aur.archlinux.org/packages/mingw-w64-cmake) itself produces a broken binary ... (at least for Wine, I don't have Windows to test) ... maybe it's also the AUR build that is broken ...

```
$ x86_64-w64-mingw32-cmake ..
...
-- CMAKE_HOST_SYSTEM_NAME="Linux"
-- CMAKE_SYSTEM_NAME="Windows"
-- CMAKE_EXECUTABLE_SUFFIX=".exe"
$ make check
...
[100%] Linking C executable test-target.exe
[100%] Built target test-target
0100:err:module:import_dll Library libgcc_s_seh-1.dll (which is needed by L"Z:\\path\\to\\libtommath\\build\\libtommath.dll") not found
0100:err:module:import_dll Library libtommath.dll (which is needed by L"Z:\\path\\to\\libtommath\\build\\test-target.exe") not found
0100:err:module:LdrInitializeThunk Importing dlls for L"Z:\\path\\to\\libtommath\\build\\test-target.exe" failed, status c0000135
make[3]: *** [CMakeFiles/check.dir/build.make:70: CMakeFiles/check] Error 53
make[2]: *** [CMakeFiles/Makefile2:137: CMakeFiles/check.dir/all] Error 2
make[1]: *** [CMakeFiles/Makefile2:144: CMakeFiles/check.dir/rule] Error 2
make: *** [Makefile:215: check] Error 2
$ file test-target*
test-target.exe: PE32+ executable (console) x86-64, for MS Windows
```

Reply:
I could build and run `test-target.exe` within a mingw docker container based on [this one](https://github.com/General-Beck/docker-mingw-qt5) I use for cross building Qt stuff for windows.

I still get an error though due to a path issue:
`make[3]: ./test-target: No such file or directory`

But manually running `test-target.exe` works [test-target.log](https://github.com/libtom/libtommath/files/8314137/test-target.log).

@sjaeckel in theory invoking `x86_64-w64-mingw32-cmake` directly should pick the right toolchain file. This is e.g. the `ccmake` output from within the build directory (NOTE the CMAKE_TOOLCHAIN_FILE variable):

```
 BUILD_SHARED_LIBS                ON
 CMAKE_BUILD_TYPE                 
 CMAKE_CXX_IMPLICIT_INCLUDE_DIR   /usr/x86_64-w64-mingw32/include
 CMAKE_C_IMPLICIT_INCLUDE_DIREC   /usr/x86_64-w64-mingw32/include
 CMAKE_GNUtoMS                    OFF
 CMAKE_INSTALL_PREFIX             /usr/x86_64-w64-mingw32
 CMAKE_TOOLCHAIN_FILE             /usr/share/mingw/toolchain-x86_64-w64-mingw32.cmake
 LSB_RELEASE                      /usr/sbin/lsb_release
```

Reply:
> I still get an error though due to a path issue:

Yeah, I've still to commit&push all my local changes ...

```
$ git diff | grep -B1 EXE
-    COMMAND ${CMAKE_CURRENT_BINARY_DIR}/test-target
+    COMMAND ${CMAKE_CURRENT_BINARY_DIR}/${LTM_TEST}${CMAKE_EXECUTABLE_SUFFIX}
```